mm2src/coins/sia/http_client.rs

```diff
@@ -0,0 +1,193 @@
+#[derive(Debug)]
+pub struct ReqwestErrorWithUrl {
+    error: reqwest::Error,
+    url: Url,
+}
+
+impl Display for ReqwestErrorWithUrl {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        write!(f, "Error: {}, URL: {}", self.error, self.url)
+    }
+}
+
+#[derive(Debug, Display)]
+pub enum SiaApiClientError {
+    Timeout(String),
+    BuildError(String),
+    ApiUnreachable(String),
+    ReqwestError(ReqwestErrorWithUrl),
+    UrlParse(url::ParseError),
+}
+
+impl From<SiaApiClientError> for String {
+    fn from(e: SiaApiClientError) -> Self { format!("{:?}", e) }
+}
+
+async fn fetch_and_parse<T: DeserializeOwned>(client: &Client, url: Url) -> Result<T, SiaApiClientError> {
+    client
+        .get(url.clone())
+        .send()
+        .await
+        .map_err(|e| {
+            SiaApiClientError::ReqwestError(ReqwestErrorWithUrl {
+                error: e,
+                url: url.clone(),
+            })
+        })?
+        .json::<T>()
+        .await
+        .map_err(|e| SiaApiClientError::ReqwestError(ReqwestErrorWithUrl { error: e, url }))
+}
+
+// https://github.com/SiaFoundation/core/blob/4e46803f702891e7a83a415b7fcd7543b13e715e/types/types.go#L181
+#[derive(Deserialize, Serialize, Debug)]
+pub struct GetConsensusTipResponse {
+    pub height: u64,
+    pub id: String, // TODO this can match "BlockID" type
+}
+
+// https://github.com/SiaFoundation/walletd/blob/9574e69ff0bf84de1235b68e78db2a41d5e27516/api/api.go#L36
+// https://github.com/SiaFoundation/walletd/blob/9574e69ff0bf84de1235b68e78db2a41d5e27516/wallet/wallet.go#L25
+#[derive(Deserialize, Serialize, Debug)]
+pub struct GetAddressesBalanceResponse {
+    pub siacoins: MmNumber,
+    #[serde(rename = "immatureSiacoins")]
+    pub immature_siacoins: MmNumber,
+    pub siafunds: u64,
+}
+
+impl SiaApiClientImpl {
+    fn new(base_url: Url, password: &str) -> Result<Self, SiaApiClientError> {
+        let mut headers = HeaderMap::new();
+        let auth_value = format!("Basic {}", base64::encode(&format!(":{}", password)));
+        headers.insert(
+            AUTHORIZATION,
+            HeaderValue::from_str(&auth_value).map_err(|e| SiaApiClientError::BuildError(e.to_string()))?,
+        );
+
+        let client = reqwest::Client::builder()
+            .default_headers(headers)
+            .timeout(Duration::from_secs(10)) // TODO make this configurable
+            .build()
+            .map_err(|e| {
+                SiaApiClientError::ReqwestError(ReqwestErrorWithUrl {
+                    error: e,
+                    url: base_url.clone(),
+                })
+            })?;
+        Ok(SiaApiClientImpl { client, base_url })
+    }
+
+    pub async fn get_consensus_tip(&self) -> Result<GetConsensusTipResponse, SiaApiClientError> {
+        let base_url = self.base_url.clone();
+        let endpoint_url = base_url
+            .join("api/consensus/tip")
```

Comment:
It's good to have these string paths such as `api/consensus/tip` as `const`

Reply:
I did as you suggested for this particular instance. I'll refactor the dynamic endpoint URLs in the next iteration.